### ISSUE TITLE
ThreadExecutor doesn't survive fork

### ISSUE THREAD

**Issue by matthewd** [ruby-concurrency/concurrent-ruby#500]
```ruby
require 'concurrent'
20.times { Concurrent::Future.execute { 1 } }
fork { Concurrent::Future.execute { 7 }.wait! }
```

This causes a deadlock in the child, because its executor is full of dead threads.

It feels like the executor should be able to detect that condition, and flush the dead workers out of the pool.

**Comment by pitr-ch** [comment]
@matthewd Thanks for contacting us and letting us know. Unfortunately this will require larger discussion. It's not just about how to add a callback to restart the threads after forking. Concurrent-ruby is designed to run on any Ruby implementation. Forking is not supported on all of them, so we need to come up with some solution/API which will work on all Ruby implementations. @matthewd could you elaborate on your use-case? It'll give us a better idea what needs to be added.

**Comment by jdantonio** [comment]
My initial reaction to this request is that this isn't a bug. Passing a thread pool to a forked process isn't something we should be reasonably expected to support. The workers we would be flushing aren't "dead" threads, they are references to threads that *don't exist* in the forked process. Forking and threading are both concurrency techniques. They have distinctly different advantages and disadvantages. Mixing the two feels like an anti-pattern to me.

That being said, I'm certainly open to discussing it further. Please describe the use case in more detail. That context will help us better understand why you are requesting this feature.

**Comment by matthewd** [comment]
concurrent-ruby has global executors (like the one used for futures). If those globals don't survive a fork, then concurrent-ruby is incompatible with forking, and that's gonna be a problem.

> Forking is not supported on all of them, so we need to come up with some solution/API which will work on all Ruby implementations.

This part is easy: if the implementation doesn't support forking, then there's no problem to worry about.

> The workers we would be flushing aren't "dead" threads, they are references to threads that don't exist in the forked process.

Indeed... I was talking about the ruby-level Thread's state.

---

My argument is that if the executor is idle at the time of the fork, the fact the threads technically still exist is, as far as the caller is concerned, merely a transparent performance optimisation.

If it's not supported to create (and then wait for) a future within a method, because somewhere much later the application may have reason to fork (and would then have an at best incomplete, at worst useless, executor pool), then I don't see how any library can use concurrent-ruby internally.

---

My proposed solution doesn't require noticing the fork at the time it occurs: you just need to know it's happened next time you go to enqueue work. In the simple case, the executor knows it was idle, and it can throw away the "dead" threads with impunity.

If you want to be paranoid (moreso than I personally think is within your remit, but not unreasonably so), you could have the executor raise if, when enqueuing work, it notices that it had work in flight and all its threads are now dead. This would make it difficult for a library to use concurrent-ruby to spawn work and then return, but at least fully-support situations where the work was completed before returning control to the calling application.

To me, the most MRI-consistent approach would be to declare any fallout from additional threads while forking to be the forking application's problem, and just internally recover from the situation the same as it would if the pools were managed by a separate supervisor thread with its own locals: throw everything away (including anything queued but not yet in progress: its caller asked for it to be executed once, and the fork parent will take care of that) and start again.

**Comment by matthewd** [comment]
Specifically, here's line currently causing a problem: https://github.com/rails/sprockets/blob/a7b20d571012318de6271290fae6c96c06d86635/lib/sprockets/manifest.rb#L168

If `compile` is called on a non-trivial manifest, the global executor gets fully populated. It the process later forks, then calls `compile` again (or uses Future for any other reason, without an explicit executor) the Future never runs, and you get a deadlock.

For something small and infrequently called like this, we can solve it by spawning a new executor inside the method, instead of allowing c-r to use its potentially-broken default. But for something hot enough to benefit from the shared pool, or with a large enough surface area that it's harder to consistently audit everything that might use the global executors... :confused:

**Comment by jdantonio** [comment]
Let me see if I can sum this up:

* If an MRI fork occurs *after* a thread pool has created one or more threads the copy of the thread pool becomes effectively useless. The current implementation assumes all threads are viable when in situation they are not.
* When a fork occurs the thread pool on the original process will continue to process the jobs in its queue.
* That's OK with us. Our goal isn't to split enqueued work between the two processes. Our concern is to ensure that the thread pool on the new process remains viable for *new* work.
* So it's acceptable for the thread pool on the new process to clear its queue of any pending jobs because we assume that they will be (or have been) handled by the original process.
* It's also OK for the thread pool in the new process to clear its collection of thread objects. They are useless anyway.
* Presumably a job that is being handled by a worker *at the moment* the fork occurs is *not* an issue. We'll want to test this but I would assume that MRI doesn't try to start a new copy of these threads on the new process (since this is the root of the issue).
* This is a non-issue if the fork occurs *before* a thread pool creates any threads of if it becomes completely idle (minimum threads of zero, all threads pruned, nothing in queue). There are no thread objects to fork.
* This is a non-issue on JRuby since JRuby [doesn't implement fork()](https://github.com/jruby/jruby/wiki/DifferencesBetweenMriAndJruby#fork-is-not-implemented). Which is good since on JRuby we use the thread pools from java.util.concurrent.
* So the issue is limited to our pure-Ruby thread pools.

The [suggested solution](https://gitter.im/ruby-concurrency/concurrent-ruby?at=56a2a7b7dc33b33c75475ab2) of checking `$$` when a job is enqueued could be implemented with a simple check [here](https://github.com/ruby-concurrency/concurrent-ruby/blob/master/lib/concurrent/executor/ruby_thread_pool_executor.rb#L146). On instantiation the executor object stores the value of `$$` then checks it first within `ns_execute`. If the value has changed it clears the queue, clears the thread pool, resets its internal "have I forked?" variable, then proceeds as normal. Since this method is called from within a mutex lock it is safe to do all of this. Potentially we could update the `ns_prune_pool` method to look for dead threads as well, and purge them.

@matthewd Does this accurately describe the issue and your suggested solution?

@chrisseaton @pitr-ch Do either of you have any concerns about this?

To me this sounds like a perfectly reasonable update but there is certainly room for discussion, especially if I am missing something.

**Comment by matthewd** [comment]
:+1:

The only very special circumstance I can foresee would be if a thread belonging to the executor were the one to fork: then its thread would still be alive and running. As the thread is now the new process's "main" thread, it seems fine (and even actively desirable) that the described logic will have removed it from the pool assuming it was dead [or more strictly, and more still true: useless to the executor].

If it forks with a block, that's fine: the child process will terminate when the end of the block is reached, and everyone will be happy. If it performs a bare fork, however, it would be possible for this no-longer-in-pool thread to hit its work loop.

I think that edge case can be saved by a check just before `ready_worker`: if $$ has changed since the thread first spawned and this thread survived, or more directly, if Thread.current == Thread.main, then it should act like a `:stop` -- the pop can only lead to deadlock.


The executor probably needs to do the pid-change check somewhere around `ns_shutdown_execution` (and maybe `ns_kill_execution`, pending its TODOs) too: if it's never been given new work, it could find itself waiting for threads that will never end.

**Comment by matthewd** [comment]
I now observe that my original example didn't wait for the Futures to complete before it forked, which could imply I was trying to spread their work across the two processes -- which would indeed be a (rather questionable) feature request.

To make that clearer for any future readers, my problem is more specifically this:

```ruby
require 'concurrent'
20.times.map { Concurrent::Future.execute { 1 } }.each(&:wait!)
# Executor is idle
fork { Concurrent::Future.execute { 7 }.wait! } # => child deadlocks
```

**Comment by pitr-ch** [comment]
@mastfish So if I understand correctly, it's not that you are forking the application (e.g. because there is a cpu bound computations you would like to spread to more MRI processes) but rather if something else forks for whatever reason you would like to be sure that you can use c-r at it won't break. Correct?

I agree c-r should fix this issue or at least raise exceptions so it does not lead deadlocks and ugly surprises. However even if this is fixed c-r will still be incompatible with forking and we should print a warning. We should then have a think about other abstractions considering what might break. We could also consider extend Kernel.fork and ensure there programatically that nothing broke, forbid forking of executor threads etc. 

> > Forking is not supported on all of them, so we need to come up with some solution/API which will work on all Ruby implementations.

> This part is easy: if the implementation doesn't support forking, then there's no problem to worry about.

I rather meant that we should support the need (assuming it's spreading cpu bound computation over more cores) for forking with our API which will do something meaningful on all implementations. On JRuby and similar, it would do nothing since the implementation is already using all cores. On MRI, it would have to somehow allow to spread the load over more processes - not an easy task though since block cannot by marshalled. Hm as I read it after myself that sounds unrealistic.

The $$ check before every job could be expensive, it has to be checked first, if it's just a volatile read or even a system call we need something else. It cannot be called for every job. I plan rework the executor to add queue per thread to avoid synchronisation as much as possible on job enqueuing, thread would steel jobs from each other when its queue is empty.

**Comment by matthewd** [comment]
> However even if this is fixed c-r will still be incompatible with forking and we should print a warning.

Really? That's a deal-breaker for me. :worried: 

> The $$ check before every job could be expensive

Possibly cheaper: remember Thread.main when the executor is built, and then check whether that and the first thread in the pool are both alive -- if one or both are dead, we've forked.

And if it helps further, I think you could maybe further avoid synchronisation by focusing on whether the chosen worker thread is still alive.

**Comment by pitr-ch** [comment]
> Really? That's a deal-breaker for me.

Maybe I formulated it with too strong words :smile:, it's rather that we do not know at the moment it probably works but we have to make sure first that there are no issues.

**Comment by pitr-ch** [comment]
I think we are in agreement that this needs to be fixed, @matthewd thanks for the suggestions, we'll try to benchmark what's the fastest way to check it an we'll add something into the MRI implementation of executor.

**Comment by jdantonio** [comment]
A few thoughts (in no particular order) and some additional information:

* I still consider threading+forking to be an anti-pattern, and possibly an edge case. My main concern is that ThreadPoolExecutor survive and behave in a predictable manner. I'd like to limit our changes to only that issue.
* I'm not concerned with any of our high-level abstractions right now. If our thread pools survive a fork with the behavior described above then all of our high-level abstractions will perform in a predictable manner. Handling threading+forking can then be an application concern (at least until *specific* issues are reported).
* Presumably this is an issue for *every* Ruby concurrency library since forking would kill their threads, too. We can look at other libraries but I doubt that any of them have done more than simply restart threads.
* I'm strongly opposed to extending/monkey-patching Kernel.fork. As a foundational library it's critical that we don't make changes to the Ruby runtime that negatively effect other gems. That's basically the issue we're facing right now: forking the parent process may cause downstream gems to behave in an unpredictable way because of us. We don't "fix" this issue by causing other downstream problems.
* I'm not concerned about the performance implications. We're focussed almost exclusively on MRI here. It's not go or C or rust, or even Java or C#. MRI was never intended to be a high-performance languages. MRI has a GIL so the main use case for c-r on that platform is concurrent I/O. A small performance hit when we post to a thread pool is likely going to be insignificant when compared to the benefit of not blocking in I/O.
* We definitely want to optimize our thread pools as much as possible in future versions, but today we need to focus on correctness. I'd like to avoid extensive refactoring of TPE if it can be avoided.
* Since our ns_execute function already does a bunch of stuff--all within a mutex--adding an additional check probably won't make that much of an impact. We definitely need to benchmark it, but the data should drive our decision. Until we implement something and benchmark it we're just guessing.
* AKAICT a read from `$$` maps to a C `getpid` call. It's a little convoluted, but the relevant parts are:
  - `$$` is implemented using `rb_define_virtual_variable ` [here](http://rxr.whitequark.org/mri/source/process.c#6645).
  - `rb_define_virtual_variable` is described [here](http://rxr.whitequark.org/mri/source/README.EXT#1127) as "Defines a virtual variable, whose behavior is defined by a pair of C functions.  The getter function is called when the variable is referenced."
  - The aforementioned definition maps `$$` to the Ruby function `get_pid`
  - The Ruby function `get_pid` in turn calls the C function `getpid` function [here](http://rxr.whitequark.org/mri/source/process.c#204).
  - This seems odd to me since the process ID should not change unless we fork so `&&` could be set once when the VM initializes, but that doesn't seem to be the case.

So my suggestion is that we do this:

* Update `ns_execute` to check `$$` and benchmark the difference.
  - If the impact is minimal then we go with it.
  - Otherwise we find another, simple comparison and use that instead.
* (Optional) Update `ns_prune_pool` to also purge dead threads (we did this in an earlier implementation of TPE).
* Resist further optimization to PTE and limit our scope to fixing the reported issue.
* Update the documentation of TPE to describe this new behavior and commit to it as part of our safety guarantees.
* Leave the high-level abstractions alone and don't say anything else in the documentation. These are esoteric details that most users won't care about.
* Resume discussions regarding other threading+forking behavior as part of the 2.0 release.
* :shipit: 

Thoughts, comments, suggestions, concerns, objections?

**Comment by thedarkone** [comment]
> I'm strongly opposed to extending/monkey-patching Kernel.fork.

My first goto solution would be to just monkey patch `Kernel#fork` :bowtie:.

----

On a serious note, the initial mechanism (the `$$` check or smth else) could be Ruby VM specific (noops for JRuby etc.) and as @eregon pointed out on gitter, in the future we could be rescued by https://bugs.ruby-lang.org/issues/5446. Maybe with a little help MRI core team could land this reasonably soon (2.4)?

**Comment by jdantonio** [comment]
I created a couple of test scripts in PR #501.

**Tl;DR:** We cannot use `Thread.main` to determine if our process has forked because Ruby copies all attributes of `Thread.main` on the new fork. However checking `$$` on every post is only 1.08x slower which is probably an acceptable performance hit.

@matthewd @pitr-ch If you have some time, please run the scripts in the aforementioned PR, make sure my logic/code is sound, and verify that you get similar results.

**Comment by eregon** [comment]
Just two notes:
* `getpid` on most OSs does not do a real syscall, but is just a read from a well-known read-only memory location.
* The Ruby Thread which is forked is the exact same instance in the child, which I would call unfortunate but it's the way it is.

**Comment by jdantonio** [comment]
I've made another update and done more testing in #501.

**Comment by schneems** [comment]
Thanks for the work here, it took me a few hours to find the original issue before we figured out tests in https://github.com/rails/sprockets-rails were forking and causing the executor threads to be dead.

**Comment by jdantonio** [comment]
@schneems Thank you for reporting it! This is an edge case we never considered. c-r will be stronger because if this fix.

**Comment by jdantonio** [comment]
Fix in PR #501 has been merged. Will release v1.0.1 soon.

**Comment by schneems** [comment]
BTW not sure if you got around to writing integration tests yet. I made a library that might help https://github.com/schneems/wait_for_it.
